File: agentscope-service/aistio/sdk/python/aistio/events.py
# Copyright 2024-2026 the original author or authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Level-2 session event model (mirrors ``asdp.SessionEventMsg``) and the
Level-3 full-history page (mirrors ``prober.MessagePage``).

事件流是统一会话页面的持久化事实源，因此保留完整消息和工具内容，不依赖可选的
Level 3 ``message-query`` 能力。
"""
from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any, List, Optional

from ._util import now_ms, rfc3339
from .proto import asdp_pb2

# ─── Event type vocabulary (aligned with asdp.proto SessionEventMsg.event_type) ───
EVENT_SESSION_START = "session_start"
EVENT_MESSAGE = "message"
EVENT_TOOL_CALL = "tool_call"
EVENT_TOOL_RESULT = "tool_result"
EVENT_SESSION_END = "session_end"
EVENT_COMPACTION = "compaction"

EVENT_TYPES = frozenset(
    {
        EVENT_SESSION_START,
        EVENT_MESSAGE,
        EVENT_TOOL_CALL,
        EVENT_TOOL_RESULT,
        EVENT_SESSION_END,
        EVENT_COMPACTION,
    }
)

# ─── Role vocabulary ───
ROLE_USER = "user"
ROLE_ASSISTANT = "assistant"
ROLE_SYSTEM = "system"
ROLE_TOOL = "tool"


def _json_bytes(value: Any) -> bytes:
    """Best-effort canonical JSON encoding; returns ``b""`` for ``None``."""
    if value is None:
        return b""
    if isinstance(value, bytes):
        data = value
    elif isinstance(value, str):
        data = value.encode("utf-8")
    else:
        data = json.dumps(value, ensure_ascii=False, separators=(",", ":")).encode("utf-8")
    return data


@dataclass
class SessionEvent:
    """One Level-2 session event.

    ``seq`` 在会话内单调递增，是控制面幂等去重键（``(session_fk, seq)`` 唯一）。
    """

    session_id: str
    seq: int
    event_type: str
    occurred_at: int = 0  # unix ms
    role: str = ""
    content: str = ""  # 完整内容；控制面事件日志是会话历史事实源
    tool_name: str = ""
    tool_input: Optional[bytes] = None  # 完整 JSON bytes
    tool_output: str = ""  # 完整输出
    tokens_in: int = 0
    tokens_out: int = 0
    duration_ms: int = 0
    framework_meta: Optional[bytes] = None  # 框架私有 JSON

    def __post_init__(self) -> None:
        if self.occurred_at <= 0:
            self.occurred_at = now_ms()

    @staticmethod
    def encode_tool_input(value: Any) -> bytes:
        """Encode a complete tool-input payload as JSON bytes."""
        return _json_bytes(value)

    @staticmethod
    def encode_meta(meta: Any) -> bytes:
        """Encode framework-private metadata as JSON bytes."""
        return _json_bytes(meta)

    def to_proto(self) -> "asdp_pb2.SessionEventMsg":
        return asdp_pb2.SessionEventMsg(
            session_id=self.session_id,
            seq=self.seq,
            event_type=self.event_type,
            occurred_at=self.occurred_at,
            role=self.role,
            content=self.content,
            tool_name=self.tool_name,
            tool_input=self.tool_input or b"",
            tool_output=self.tool_output,
            tokens_in=self.tokens_in,
            tokens_out=self.tokens_out,
            duration_ms=self.duration_ms,
            framework_meta=self.framework_meta or b"",
        )


@dataclass
class MessageItem:
    """One full-content history entry (Level 3; mirrors ``prober.MessageItem``)."""

    seq: int
    role: str
    content: str
    tool_name: str = ""
    tool_input: Any = None  # JSON-able
    tool_output: str = ""
    occurred_at: int = 0  # unix ms

    def to_json_dict(self) -> dict:
        d: dict = {"seq": self.seq, "role": self.role, "content": self.content}
        if self.tool_name:
            d["toolName"] = self.tool_name
        if self.tool_input is not None:
            d["toolInput"] = self.tool_input
        if self.tool_output:
            d["toolOutput"] = self.tool_output
        ts = rfc3339(self.occurred_at)
        if ts:
            d["occurredAt"] = ts
        return d


@dataclass
class MessagePage:
    """Paginated Level-3 full-history response (mirrors ``prober.MessagePage``)."""

    session_id: str
    messages: List[MessageItem] = field(default_factory=list)
    offset: int = 0
    limit: int = 50
    total: int = 0

    def __post_init__(self) -> None:
        if not self.total:
            self.total = len(self.messages)

    def to_json_dict(self) -> dict:
        return {
            "sessionId": self.session_id,
            "offset": self.offset,
            "limit": self.limit,
            "total": self.total,
            "messages": [m.to_json_dict() for m in self.messages],
        }

File: agentscope-service/aistio/internal/store/postgres/events.go
// Copyright 2024-2026 the original author or authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package postgres

import (
	"context"
	"errors"
	"fmt"
	"strings"
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"

	"github.com/spring-ai-alibaba/aistio/internal/store"
)

type eventRepo struct {
	pool *pgxpool.Pool
}

const sessionEventNotifyChannel = "aistio_session_events"

func (r *eventRepo) Append(ctx context.Context, event *store.SessionEvent) error {
	if event.OccurredAt.IsZero() {
		event.OccurredAt = time.Now().UTC()
	}
	tx, err := r.pool.Begin(ctx)
	if err != nil {
		return fmt.Errorf("postgres events begin append: %w", err)
	}
	defer func() { _ = tx.Rollback(ctx) }()
	err = tx.QueryRow(ctx, `
		INSERT INTO session_events (
			session_fk, seq, event_type, role, content, tool_name, tool_input,
			tool_output, tokens_in, tokens_out, duration_ms, framework_meta, occurred_at
		) VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13)
		ON CONFLICT (session_fk, seq) DO NOTHING
		RETURNING id`,
		event.SessionFK, event.Seq, event.EventType, nullStr(event.Role), nullStr(event.Content),
		nullStr(event.ToolName), nullJSON(event.ToolInput), nullStr(event.ToolOutput),
		nullInt(event.TokensIn), nullInt(event.TokensOut), nullInt(event.DurationMs),
		nullJSON(event.FrameworkMeta), event.OccurredAt,
	).Scan(&event.ID)
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			// (session_fk, seq) already exists — idempotent success, aligned
			// with the memory driver.
			return store.ErrConflict
		}
		return fmt.Errorf("postgres events append: %w", err)
	}
	if _, err := tx.Exec(ctx, "SELECT pg_notify('"+sessionEventNotifyChannel+"', $1)", event.SessionFK.String()); err != nil {
		return fmt.Errorf("postgres events notify: %w", err)
	}
	if err := tx.Commit(ctx); err != nil {
		return fmt.Errorf("postgres events commit: %w", err)
	}
	return nil
}

func (r *eventRepo) WaitForNew(ctx context.Context, sessionFK uuid.UUID, afterSeq int) error {
	conn, err := r.pool.Acquire(ctx)
	if err != nil {
		return fmt.Errorf("postgres events acquire listener: %w", err)
	}
	defer conn.Release()
	if _, err := conn.Exec(ctx, "LISTEN "+sessionEventNotifyChannel); err != nil {
		return fmt.Errorf("postgres events listen: %w", err)
	}
	defer func() {
		cleanupCtx, cancel := context.WithTimeout(context.Background(), time.Second)
		defer cancel()
		_, _ = conn.Exec(cleanupCtx, "UNLISTEN "+sessionEventNotifyChannel)
	}()

	// LISTEN is active before this check, closing the race between the last SSE
	// page read and starting the wait.
	var available bool
	if err := conn.QueryRow(ctx,
		"SELECT EXISTS (SELECT 1 FROM session_events WHERE session_fk=$1 AND seq>$2)",
		sessionFK, afterSeq).Scan(&available); err != nil {
		return fmt.Errorf("postgres events check listener cursor: %w", err)
	}
	if available {
		return nil
	}
	for {
		notification, err := conn.Conn().WaitForNotification(ctx)
		if err != nil {
			return err
		}
		if notification.Payload == sessionFK.String() {
			return nil
		}
	}
}

func (r *eventRepo) List(ctx context.Context, sessionFK uuid.UUID, opts ...store.EventOption) ([]*store.SessionEvent, error) {
	o := store.ResolveEventOptions(opts)
	var (
		conds = []string{"session_fk=$1"}
		args  = []any{sessionFK}
	)
	if o.EventType != "" {
		args = append(args, o.EventType)
		conds = append(conds, fmt.Sprintf("event_type=$%d", len(args)))
	}
	if o.Since != nil {
		args = append(args, *o.Since)
		conds = append(conds, fmt.Sprintf("occurred_at>=$%d", len(args)))
	}
	if o.Until != nil {
		args = append(args, *o.Until)
		conds = append(conds, fmt.Sprintf("occurred_at<=$%d", len(args)))
	}
	if o.Before != nil {
		args = append(args, *o.Before)
		conds = append(conds, fmt.Sprintf("occurred_at<$%d", len(args)))
	}
	if o.BeforeSeq != nil {
		args = append(args, *o.BeforeSeq)
		conds = append(conds, fmt.Sprintf("seq<$%d", len(args)))
	}
	if o.AfterSeq != nil {
		args = append(args, *o.AfterSeq)
		conds = append(conds, fmt.Sprintf("seq>$%d", len(args)))
	}

	where := strings.Join(conds, " AND ")
	order := "ORDER BY seq ASC"
	if o.NewestFirst && o.Limit > 0 {
		// Newest page: DESC + LIMIT, then reverse to chronological ASC for callers.
		q := `SELECT id, session_fk, seq, event_type, role, content, tool_name, tool_input,
			tool_output, tokens_in, tokens_out, duration_ms, framework_meta, occurred_at
			FROM session_events WHERE ` + where + ` ORDER BY seq DESC`
		args = append(args, o.Limit)
		q += fmt.Sprintf(" LIMIT $%d", len(args))
		out, err := r.scanEvents(ctx, q, args)
		if err != nil {
			return nil, err
		}
		for i, j := 0, len(out)-1; i < j; i, j = i+1, j-1 {
			out[i], out[j] = out[j], out[i]
		}
		return out, nil
	}

	q := `SELECT id, session_fk, seq, event_type, role, content, tool_name, tool_input,
		tool_output, tokens_in, tokens_out, duration_ms, framework_meta, occurred_at
		FROM session_events WHERE ` + where + ` ` + order
	if o.Limit > 0 {
		args = append(args, o.Limit)
		q += fmt.Sprintf(" LIMIT $%d", len(args))
	}
	if o.Offset > 0 {
		args = append(args, o.Offset)
		q += fmt.Sprintf(" OFFSET $%d", len(args))
	}
	return r.scanEvents(ctx, q, args)
}

func (r *eventRepo) scanEvents(ctx context.Context, q string, args []any) ([]*store.SessionEvent, error) {
	rows, err := r.pool.Query(ctx, q, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []*store.SessionEvent
	for rows.Next() {
		e := &store.SessionEvent{}
		var role, content, toolName, toolOutput *string
		var toolInput, meta []byte
		var tokensIn, tokensOut, durationMs *int
		if err := rows.Scan(
			&e.ID, &e.SessionFK, &e.Seq, &e.EventType, &role, &content, &toolName, &toolInput,
			&toolOutput, &tokensIn, &tokensOut, &durationMs, &meta, &e.OccurredAt,
		); err != nil {
			return nil, err
		}
		e.Role = deref(role)
		e.Content = deref(content)
		e.ToolName = deref(toolName)
		e.ToolOutput = deref(toolOutput)
		e.ToolInput = toolInput
		e.FrameworkMeta = meta
		if tokensIn != nil {
			e.TokensIn = *tokensIn
		}
		if tokensOut != nil {
			e.TokensOut = *tokensOut
		}
		if durationMs != nil {
			e.DurationMs = *durationMs
		}
		out = append(out, e)
	}
	return out, rows.Err()
}

func nullInt(n int) any {
	if n == 0 {
		return nil
	}
	return n
}
